### PR TITLE
CountVectorizerFeaturizer + TfidfVectorizerFeaturizer

### PR DESCRIPTION
**Pull request by wangyems**
todo: 
1. max_features(would take another 1 or 2 hours)
2. more tests will be added tomorrow(for now there are only for default parameters)
3. Strings related stuff

### CODE REVIEW

File: src/Featurizers/CountVectorizerFeaturizer.h [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+                    throw std::invalid_argument("min_df");
+
+                return min_df;
+            }()
+        )
+    ),
+    _max_features(std::move(max_features)),
+    _vocabulary(std::move(vocabulary)),
+    _binary(std::move(binary)) {
+}
```
[the remark this answers is not in the document]
Comment (wangyems) [reply src/Featurizers/CountVectorizerFeaturizer.h]: implemented in documentstatsestimator

File: src/Featurizers/TfidfVectorizerFeaturizer.h [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+                    ++iter_tf->second;
+                } else {
+                    TermFreqMap.insert(std::make_pair(std::make_tuple(iter_start, iter_end), 1));
+                }
+            }
+        );
+
+        std::float_t normVal = 0.0f;
+        std::vector<std::tuple<std::uint32_t, std::float_t>> results;
+        for (auto const & pair : TermFreqMap) {
```
[the remark this answers is not in the document]
Comment (wangyems) [reply src/Featurizers/TfidfVectorizerFeaturizer.h]: This is the desired behavior

File: src/Featurizers/CountVectorizerFeaturizer.h [hunk 3]
```diff
@@ -0,0 +1,347 @@
+// ----------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License
+// ----------------------------------------------------------------------
+#pragma once
```
Comment (yuslepukhin) [inline src/Featurizers/CountVectorizerFeaturizer.h]: >ragma once [](start = 2, length = 10)

Add this to a CMakefile

File: src/Featurizers/TfidfVectorizerFeaturizer.h [hunk 4]
```diff
@@ -0,0 +1,461 @@
+// ----------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License
+// ----------------------------------------------------------------------
+#pragma once
```
Comment (yuslepukhin) [inline src/Featurizers/TfidfVectorizerFeaturizer.h]: >gma once [](start = 4, length = 8)

Add this to a CMake file

File: src/Featurizers/Structs.h [hunk 5]
```diff
@@ -1,245 +1,329 @@
...
+    bool operator==(TFStruct const &other) const;
+};
+
+/////////////////////////////////////////////////////////////////////////
+///  \struct        TFIDFStruct
+///  \brief         Struct to hold return value of tfidf vectorizer,
+///                 consist of <dictid, tfidf value>
+///
+struct TFIDFStruct {
+    std::uint32_t const DictionaryId;                              // dict id
```
Comment (yuslepukhin) [inline src/Featurizers/Structs.h]: >DictionaryId [](start = 24, length = 12)

Should not this have underscore as a member?

File: src/Featurizers/Structs.h [hunk 6]
```diff
@@ -1,245 +1,329 @@
...
+    bool operator!=(SingleValueSparseVectorEncoding const &other) const;
+};
+
+/////////////////////////////////////////////////////////////////////////
+///  \struct        TFStruct
+///  \brief         Struct to hold return value of count vectorizer,
+///                 consist of <dictid, # of appearances>
+///
+struct TFStruct {
+    std::uint32_t const DictionaryId;                              // dict id
```
Comment (yuslepukhin) [inline src/Featurizers/Structs.h]: >DictionaryId [](start = 24, length = 12)

Underscore in the name?

File: src/Featurizers/Structs.h [hunk 7]
```diff
@@ -1,245 +1,329 @@
...
+namespace Featurizers {
+
+/////////////////////////////////////////////////////////////////////////
+///  \class         SparseVectorEncoding
+///  \brief         Object that contains information that can be used to
+///                 create a sparse vector without creating the vector
+///                 itself.
+///
+struct OneHotStruct {
+    std::uint32_t const Index;                                    // category index
```
Comment (yuslepukhin) [inline src/Featurizers/Structs.h]: >Index [](start = 24, length = 5)

Member names underscore?

File: src/Featurizers/Structs.h [hunk 8]
```diff
@@ -1,245 +1,329 @@
...
+// ----------------------------------------------------------------------
+
+// ----------------------------------------------------------------------
+// |
+// |  SparseVectorEncoding::ValueEncoding
+// |
+// ----------------------------------------------------------------------
+template <typename T>
+SparseVectorEncoding<T>::ValueEncoding::ValueEncoding(value_type value, std::uint64_t index) :
+    Value(std::move(value)),
```
Comment (yuslepukhin) [inline src/Featurizers/Structs.h]: >std::move [](start = 10, length = 9)

Why?

File: src/Featurizers/Structs.h [hunk 9]
```diff
@@ -1,245 +1,329 @@
...
+            }()
+        )
+    ),
+    Values(
+        std::move(
+            [&values](void) -> std::vector<ValueEncoding> & {
+                if(values.empty())
+                    throw std::invalid_argument("'values' is empty");
+
+                // Ensure that the values are ordered by index
```
Comment (yuslepukhin) [inline src/Featurizers/Structs.h]: >// Ensure that the values are ordered by index [](start = 16, length = 46)

std::is_sorted?

File: src/Featurizers/Structs.h [hunk 10]
```diff
@@ -1,245 +1,329 @@
...
+        && Index == other.Index;
+}
+
+#if (defined __clang__)
+#   pragma clang diagnostic pop
+#endif
+
+template <typename T>
+bool SingleValueSparseVectorEncoding<T>::operator!=(SingleValueSparseVectorEncoding const &other) const {
+    return (*this == other) == false;
```
Comment (yuslepukhin) [inline src/Featurizers/Structs.h]: > == false; [](start = 27, length = 10)

You compare bool to another bool to get a bool. You have so many bools!
return !(*this == other);

File: src/Featurizers/TfidfVectorizerFeaturizer.h [hunk 11]
```diff
@@ -0,0 +1,461 @@
...
+    void execute_impl(typename BaseType::InputType const &input, typename BaseType::CallbackFunction const &callback) override {
+
+        typename std::map<IterRangeType, std::uint32_t, Components::IterRangeComp> TermFreqMap;
+
+        //todo: will use vector<functor> after string header file is done
+        Components::split_temp( 
+            input, 
+            [&TermFreqMap] (std::string::const_iterator & iter_start, std::string::const_iterator & iter_end) {
+                
+                typename std::map<IterRangeType, std::uint32_t, Components::IterRangeComp>::iterator iter_tf(TermFreqMap.find(std::make_tuple(iter_start, iter_end)));
```
Comment (yuslepukhin) [inline src/Featurizers/TfidfVectorizerFeaturizer.h]: >typename std::map<IterRangeType, std::uint32_t, Components::IterRangeComp>::iterator iter_tf [](start = 16, length = 92)

Would it be possible to use auto?

File: src/Featurizers/TfidfVectorizerFeaturizer.h [hunk 12]
```diff
@@ -0,0 +1,461 @@
...
+    bool const                               SublinearTf;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    TfidfVectorizerTransformer(IndexMapType labels, IndexMapType docufreq, std::uint32_t totalnumdocus, bool binary, 
+                               std::string norm, bool use_idf, bool smooth_idf, bool sublinear_tf);
+    TfidfVectorizerTransformer(Archive &ar);
```
Comment (yuslepukhin) [inline src/Featurizers/TfidfVectorizerFeaturizer.h]: > Tf [](start = 3, length = 3)

explicit

File: src/Featurizers/TfidfVectorizerFeaturizer.h [hunk 13]
```diff
@@ -0,0 +1,461 @@
...
+
+        //todo: will use vector<functor> after string header file is done
+        Components::split_temp( 
+            input, 
+            [&TermFreqMap] (std::string::const_iterator & iter_start, std::string::const_iterator & iter_end) {
+                
+                typename std::map<IterRangeType, std::uint32_t, Components::IterRangeComp>::iterator iter_tf(TermFreqMap.find(std::make_tuple(iter_start, iter_end)));
+                
+                if (iter_tf != TermFreqMap.end()) {
+                    ++iter_tf->second;
```
Comment (yuslepukhin) [inline src/Featurizers/TfidfVectorizerFeaturizer.h]: > ++iter_tf->second; [](start = 19, length = 19)

Is second represents end iterator? You can't do anything with end iterators except comparision

File: src/Featurizers/TfidfVectorizerFeaturizer.h [hunk 14]
```diff
@@ -0,0 +1,461 @@
...
+                std::float_t tf;
+                std::float_t idf;
+
+                //calculate tf
+                if (Binary) {
+                    tf = 1.0f;
+                } else if (!SublinearTf) {
+                    tf = static_cast<std::float_t>(pair.second);
+                } else {
+                    tf = 1.0f + static_cast<std::float_t>(log(pair.second));
```
Comment (yuslepukhin) [inline src/Featurizers/TfidfVectorizerFeaturizer.h]: >static_cast<std::float_t>(log(pair.second)); [](start = 32, length = 44)

use std::log overload for float and avoid casting. Every time you want to cast, ask yourself if you can avoid it.

File: src/Featurizers/TfidfVectorizerFeaturizer.h [hunk 15]
```diff
@@ -0,0 +1,461 @@
...
+            if (iter_label != Labels.end()) {
+                
+                std::float_t tf;
+                std::float_t idf;
+
+                //calculate tf
+                if (Binary) {
+                    tf = 1.0f;
+                } else if (!SublinearTf) {
+                    tf = static_cast<std::float_t>(pair.second);
```
Comment (yuslepukhin) [inline src/Featurizers/TfidfVectorizerFeaturizer.h]: >static_cast<std::float_t> [](start = 25, length = 25)

Why is this necessary?

File: src/Featurizers/TfidfVectorizerFeaturizer.h [hunk 16]
```diff
@@ -0,0 +1,461 @@
...
+            }
+        }
+
+        // l2-norm calibration
+        if (Norm == "l2") {
+            normVal = sqrt(normVal);
+        }   
+
+        for (auto & result : results) {
+            callback(TFIDFStruct(std::get<0>(result), std::get<1>(result) / normVal));
```
[the remark this answers is not in the document]
Comment (yuslepukhin) [reply src/Featurizers/TfidfVectorizerFeaturizer.h]: Is normVal a floating point?

---
In reply to: [350855534](https://github.com/microsoft/FeaturizersLibrary/pull/7#discussion_r350855534) [](ancestors = 350855534)

File: src/Featurizers/TfidfVectorizerFeaturizer.h [hunk 17]
```diff
@@ -0,0 +1,461 @@
...
+                } else {
+                    TermFreqMap.insert(std::make_pair(std::make_tuple(iter_start, iter_end), 1));
+                }
+            }
+        );
+
+        std::float_t normVal = 0.0f;
+        std::vector<std::tuple<std::uint32_t, std::float_t>> results;
+        for (auto const & pair : TermFreqMap) {
+            std::string const word = std::string(std::get<0>(pair.first), std::get<1>(pair.first));
```
Comment (yuslepukhin) [inline src/Featurizers/TfidfVectorizerFeaturizer.h]: >std::string(std::get<0>(pair.first), std::get<1>(pair.first)); [](start = 37, length = 62)

Not clear what it does. It looks like you are putting binary data to a std::string. It can do it, is it what you want?

File: src/Featurizers/CountVectorizerFeaturizer.h [hunk 18]
```diff
@@ -0,0 +1,347 @@
...
+    IndexMapType const                       Labels;
+    bool const                               Binary;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    CountVectorizerTransformer(IndexMapType map, bool binary);
+    CountVectorizerTransformer(Archive &ar);
```
Comment (yuslepukhin) [inline src/Featurizers/CountVectorizerFeaturizer.h]: > Cou [](start = 3, length = 4)

explicit

File: src/Featurizers/CountVectorizerFeaturizer.h (hunk 3 above)
Comment (wangyems) [reply src/Featurizers/CountVectorizerFeaturizer.h]: Will do in future PRs

File: src/Featurizers/TfidfVectorizerFeaturizer.h (hunk 4 above)
Comment (wangyems) [reply src/Featurizers/TfidfVectorizerFeaturizer.h]: Will do in future PRs

File: src/Featurizers/Structs.h (hunk 5 above)
Comment (wangyems) [reply src/Featurizers/Structs.h]: This struct will be deprecated in next update

File: src/Featurizers/Structs.h (hunk 6 above)
Comment (wangyems) [reply src/Featurizers/Structs.h]: This struct will be deprecated in next update

File: src/Featurizers/Structs.h (hunk 7 above)
Comment (wangyems) [reply src/Featurizers/Structs.h]: the same
